Added support for redis database selection(Fixes #1609)

ashishk-1 opened the pull request:
"ModPagespeedRedisDatabaseSelectIndex index" property added in conf file.
When enabled, selects the redis database mentioned by index.

- After redis database connection, executing a redis command to select database. 
- In redis_cache_tests, select command is added in HandleClientConnection before get request. This tests the mock redis socket connection and get request.
- Default redis database index is 0.

jmarantz commented on html/doc/system.html:
```diff
@@ -475,6 +475,18 @@ <h4 id="redis">Configuring Redis</h4>
+    </p>
+<dl>
+  <dt>Apache:<dd><pre class="prettyprint"
+     >ModPagespeedRedisDatabaseSelectIndex index</pre>
```
Is this a DatabaseSelectIndex, or are we selecting a DatabaseIndex?  Just want to make sure we get a name here that makes sense to Redis people.  If it's really called a DatabaseSelectIndex then this is the right spelling, and we should also change the param below from "int database_index" to "int database_select_index". 

  If we are selecting a DatabaseIndex then we should remove the "Select" from the name.

ashishk-1 replied:
We are selecing database index. So modifying property name to ModPagespeedRedisDatabaseIndex

jmarantz commented on pagespeed/system/redis_cache.cc:
```diff
@@ -599,6 +619,18 @@ bool RedisCache::Connection::EnsureConnection() {
 }
 
+bool RedisCache::Connection::EnsureDatabaseSelection() {
+  RedisReply reply = RedisCommand(StrCat("SELECT ",
```
If this is incompatible with clustering, then maybe we should conditionalize this on options->has_database_index().

Even if it's not incompatible with clustering, should we just omit the explicit db index selection if not specified in the config?  I'm not enough of a redis expert to know the right thing here, but if it's working today maybe we shouldn't bother sending the command unless explicitly configured.

ashishk-1 replied:
Added method has_redis_database_index() to check property from config. Selecting database only if property present in config.

jmarantz commented on pagespeed/system/redis_cache_test.cc:
```diff
@@ -45,6 +45,7 @@ namespace net_instaweb {
 namespace {
   static const int kReconnectionDelayMs = 10;
   static const int kTimeoutUs = 100 * Timer::kMsUs;
+  static const int kDatabaseIndex = 0;
```
It's fine to have a constant index for the indexing tests.  But can you add a test that opens 2 clients with the same redis server, different DB indexes.  Make sure a given key in one client is independent of the same key in the other one.

jmarantz commented on pagespeed/system/system_rewrite_options.h:
```diff
@@ -187,6 +188,9 @@ class SystemRewriteOptions : public RewriteOptions {
   }
+  int redis_database_index() const {
+    return redis_database_index_.value();
+  }
```
add has_redis_database_index() modeled after the other has_ methods.

oschaaf commented:
Travis ran into one failing test, which looks relevant to this change:
```
[ RUN      ] RedisCacheTest.ShutDownDuringConnection
pagespeed/system/redis_cache_test.cc:433: Failure
Value of: recvSize
  Actual: 0
Expected: ((sizeof(ArraySizeHelper(kSelectRequest))) - 1)
Which is: 23
pagespeed/system/redis_cache_test.cc:435: Failure
Value of: requestBuf
  Actual: ""
Expected: kSelectRequest
Which is: "*2\r\n$6\r\nSELECT\r\n$1\r\n0\r\n"
[  FAILED  ] RedisCacheTest.ShutDownDuringConnection (1 ms)
```

ashishk-1 commented:
Added patch with the following changes:

- property name changed to ModPagespeedRedisDatabaseIndex
- added unit test for multiple databases, also rearrangement of unit test code
- selecting database only when property specified in conf file
- documentation updated

oschaaf commented on pagespeed/system/redis_cache_test.cc:
```diff
@@ -73,35 +75,40 @@ class RedisCacheTest : public CacheTestBase {
+    cache_.back()->StartUp();
+  }
+
   void InitRedisWithCustomServer() {
```
InitRedisWithCustomDatabaseIndex may be a better name

oschaaf commented on pagespeed/system/redis_cache.cc:
```diff
@@ -599,6 +619,21 @@ bool RedisCache::Connection::EnsureConnection() {
 
+bool RedisCache::Connection::EnsureDatabaseSelection() {
+  // dont select database if database index property not specified in config
+  if (database_index_ != -1) {
```
Maybe constantize -1 to REDIS_DATABASE_INDEX_UNSET ?

ashishk-1 commented:
Added review changes:
- using kRedisDatabaseIndexNotSet as -1 in redis_cache when property not mentioned in config.
- InitRedisCache method renamed to InitRedisWithCustomDatabaseIndex.

ashishk-1 commented:
Removed default argument.
